cassiopeia/data/CassiopeiaTree.py

```diff
@@ -0,0 +1,185 @@
+        """Returns internal nodes in tree.
+        """
+        if self.__network is None:
+            return None
+        return [
+            n
+            for n in self.__network
+            if self.__network.out_degree(n) > 1
+            and self.__network.in_degree(n) > 0
+        ]
+
+    @property
+    def nodes(self) -> List[str]:
+        """Retruns all nodes in tree.
+        """
+        if self.__network is None:
+            return None
+        return [n for n in self.__network]
+
+    def reconstruct_ancestral_characters(self):
+        """Reconstruct ancestral character states
+        """
+        pass
+
+    def children(self, node: str):
+        """Gets the children of a given node.
+        """
+        pass
+
+    def set_age(self, node: str, age: float) -> None:
+        """Sets the age of a node.
+        """
+        pass
+
+    def get_age(self, node: str) -> float:
+        """Gets the age of a node.
+        """
+        pass
+
+    def set_branch_length(self, parent: str, child: str, length: float) -> None:
+        """Sets the length of a branch.
+        """
+        pass
+
+    def get_branch_length(self, parent: str, child: str) -> float:
+        """Gets the length of a branch.
+        """
+        pass
+
+    def set_state(self, node: str, character: int, state: int) -> None:
+        """Sets the state of a single character for a node.
+        """
+        pass
+
+    def set_states(self, node: str, states: List[int]) -> None:
+        """Sets all the states for a particular node.
+        """
+        pass
+
+    def get_state(self, node: str, character: int) -> int:
+        """Gets the state of a single character for a particular node.
+        """
+        pass
+
+    def get_states(self, node: str) -> List[int]:
+        """Gets all the character states for a particular node.
+        """
+        pass
+
+    def depth_first_traverse_nodes(self, postorder: bool = True):
+        """Depth first traversal of the tree.
+
+        Returns the nodes from a DFS on the tree.
+
+        Args:
+            postorder: Return the nodes in postorder. If False, returns in 
+                preorder.
+        """
+        pass
+
+    def leaves_in_subtree(self, node) -> List[str]:
+        """Get leaves in subtree below a given node.
+        """
+        pass
+
+    def get_newick(self) -> str:
+        """Returns newick format of tree.
+        """
+        return utilities.to_newick(self.__network)
+
+    def get_mean_depth_of_tree(self) -> float:
+        """Computes mean depth of tree.
+        """
+        pass
+
+    def cophenetic_correlation(self) -> float:
+        """Computes cophenetic correlation.
+
+        Compares the character-state distances with the tree tree distances
+        to compute the cophenetic correlation as a measure of agreement between
+        the two distances.
+
+        Returns
+            The cophenetic correlation
+        """
+        pass
```

Comment:
Hey Matt, 
I think this is a great list of functionalities to start. I have a couple more suggestions that might be useful down the line, let me know what you think: 
- get_edge_mutations: returns the character/state mutations newly acquired between 2 nodes
- remove_and_prune: removes a leaf sample from the tree and prunes back dead lineages
- get_degree: returns the degree of a node
- get_leaf_descendants: returns the leaf descendants of a node
- colless_index: returns the colless index calculated on the tree
- get_max_depth_of_tree: or have a function that just returns the depth for each node in the tree, and then you can take whatever statistics on it you want
- relabel_nodes: takes a dictionary, relabels nodes as such
- collapse unifurcations: collapses unifurcations on the tree

Reply:
Also, maybe we should move collapse_tree here as well and rename it collapse_mutationless_edges

Reply:
Good ideas here! 

I've added functions for `get_edge_mutations`, `remove_and_prune`, `colless_index`, `get_max_depth_of_tree`, `relabel_nodes`, `collapse_mutationless_edges`, and `collapse_unifurcations`.

`get_leaf_descendants` is a function that already exists, called `leaves_in_subtree`.

Reply:
thanks!